%%
%% Copyright 2022 OXFORD UNIVERSITY PRESS
%%
%% This file is part of the 'oup-authoring-template Bundle'.
%% ---------------------------------------------
%%
%% It may be distributed under the conditions of the LaTeX Project Public
%% License, either version 1.2 of this license or (at your option) any
%% later version.  The latest version of this license is in
%%    http://www.latex-project.org/lppl.txt
%% and version 1.2 or later is part of all distributions of LaTeX
%% version 1999/12/01 or later.
%%
%% The list of all files belonging to the 'oup-authoring-template Bundle' is
%% given in the file `manifest.txt'.
%%
%% Template article for OXFORD UNIVERSITY PRESS's document class `oup-authoring-template'
%% with bibliographic references
%%

%%%CONTEMPORARY%%%
\documentclass[unnumsec,webpdf,contemporary,large]{oup-authoring-template}%
%\documentclass[unnumsec,webpdf,contemporary,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,contemporary,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,contemporary,small]{oup-authoring-template}

%%%MODERN%%%
%\documentclass[unnumsec,webpdf,modern,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,modern,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,small]{oup-authoring-template}

%%%TRADITIONAL%%%
%\documentclass[unnumsec,webpdf,traditional,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,traditional,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,namedate,webpdf,traditional,medium]{oup-authoring-template}
%\documentclass[namedate,webpdf,traditional,small]{oup-authoring-template}

%\onecolumn % for one column layouts

%\usepackage{showframe}
\usepackage{anyfontsize}
\usepackage{multirow}
\graphicspath{{doc/Images/}} % will allow Latex to find the images folder
\usepackage{graphicx}
\sloppy % allows line spacing stretch to justify the text
% line numbers
%\usepackage[mathlines, switch]{lineno}
%\usepackage[right]{lineno}


\theoremstyle{thmstyleone}%
\newtheorem{theorem}{Theorem}%  meant for continuous numbers
%%\newtheorem{theorem}{Theorem}[section]% meant for sectionwise numbers
%% optional argument [theorem] produces theorem numbering sequence instead of independent numbers for Proposition
\newtheorem{proposition}[theorem]{Proposition}%
%%\newtheorem{proposition}{Proposition}% to get separate numbers for theorem and proposition etc.
\theoremstyle{thmstyletwo}%
\newtheorem{example}{Example}%
\newtheorem{remark}{Remark}%
\theoremstyle{thmstylethree}%
\newtheorem{definition}{Definition}%
\usepackage{hyperref}%

\begin{document}


\journaltitle{Journal Title Here}
\DOI{DOI HERE}
\copyrightyear{2022}
\pubyear{2019}
\access{Advance Access Publication Date: Day Month Year}
\appnotes{Paper}

\firstpage{1}

%\subtitle{Subject Section}

\title[Short Article Title]{TersectBrowser+ : A single browser for in-depth introgression analysis of re-sequenced genome datesets}

% With Tom first, as first author and Students in alphabetical order. 
\author[1,$\ast$]{Tomasz Kurowski\ORCID{}}
\author[2]{Gabrielle Baumberg}
\author[2]{Gregory Lupton}
\author[2]{Tanya Stead}
\author[2]{David Oluwasusi}

\authormark{Author Name et al.}

\address[1]{\orgdiv{Centre for Soil, Agrifood and Biosciences}, \orgname{Cranfield University}, \orgaddress{\street{College Road}, \postcode{MK43 0AL}, \state{Cranfield}, \country{UK}}}
\address[2]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[3]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[4]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}

\corresp[$\ast$]{Corresponding author. \href{email:email-id.com}{email-id.com}}

\received{Date}{0}{Year}
\revised{Date}{0}{Year}
\accepted{Date}{0}{Year}

%\editor{Associate Editor: Name}

%\abstract{
%\textbf{Motivation:} .\\
%\textbf{Results:} .\\
%\textbf{Availability:} .\\
%\textbf{Contact:} \href{name@email.com}{name@email.com}\\
%\textbf{Supplementary information:} Supplementary data are available at \textit{Journal Name}
%online.}

% Variant comparison important, tools are slow, what Tersect does. Tersect produced because tools slow. We modify Tersect creating Tersect+ to bringing additional functionality to aid the investigation of cultivar introgressions for biologists.  

\abstract{The field of cultivar genomics has evolved rapidly, with growing emphasis on mapping introgressionsâ€”segments of DNA from wild plant species that have been incorporated into cultivated lines. A variety of bioinformatic tools aim to interpret this expanding complexity of plant genomic data, predict introgressions, and identify commercially relevant regions for future breeding. Demand is increasing for tailored tools to interactively explore these complex genomic queries in real time. Created in 2022, the Web-based Tersect Browser can dynamically display introgression patterns by applying flexible set theoretical expressions to sets of sequence variant data. Our 2025 development of Tersect Browser offers multiple front-end extension capabilities so that the user can view and analyze data without leaving the page. We identify previously published introgressions with our tool, illustrating its potential for future research, and enriching plant breeders' understanding of crop cultivar genomics.}
\keywords{introgression, browser, resequenced, SNP}

% \boxedtext{
% \begin{itemize}
% \item Key boxed text here.
% \item Key boxed text here.
% \item Key boxed text here.
% \end{itemize}}

\maketitle
\section{Introduction}

% # Intro / Current Situation & Problem & Tersect Solution
% Large-scale resequencing projects enable comparative genomic studies to identify key variants and structures like haplotypes or introgressions. studies rely on set theoretical operations, but current tools (e.g., BEDOPS, BCFtools, BEDTools) are limited in query complexity and speed. Tersect, using its own query language and fast execution through bitmap indices, enabling complex, real-time genome queries. This will become: a part on slowness 'As shown in [ref = Tom Fady paper] tools slow'. Then add something on functionality of these tools EITHER 'slow tools have functionality' or 'tools do not have functionality'.  Then 'Tersect+ provides functionality. 

Cultivar genomics has evolved rapidly, with growing emphasis on mapping introgressions, segments of DNA from wild plant species that have been incorporated into cultivated lines. A variety of bioinformatic tools aim to interpret this expanding complexity of plant genomic data, predict introgressions, and identify commercially relevant regions for future breeding (Qin et al., 2021) \cite{Qinetal2023}. Demand for tailored tools to explore these complex genomic queries in real-time is increasing. 

As shown previously by Kurowski & Mohareb (2020)\cite{KurowskiMohareb2019}, Tersect is a lightweight, high-performing, real-time genome querying command-line utility that uses its own query language and bitmap indices to interpret and apply flexible set theoretical expressions to sets of sequence variant data for tomato genomes. In comparison to other tools such as BEDOPS, BCFtools and BDETools (Quinlan & Hall, 2010\cite{Qinlan2010}; Danecek et al., 2011)\cite{}, Tersect performs from three to over a hundred times faster than BCFTools (Kurowski & Mohareb, 2020)\cite{KurowskiMohareb2019}. 

For biologists, command-line-based tools can be cumbersome. Web-app based tools with intuitive data visualization, real-time querying, and introgression mapping provide much better user experiences and better meet the requirements for interactive exploration of large-scale variant datasets and immediate visual feedback on genomic regions of interest.  

In 2022, a web-based version of Tersect, Tersect Browser, was created (Kurowski, 2023\cite{KurowskiphdThesis}) using \boldsymbol{Include dataset}. Tersect Browser offered multiple front-end extension capabilities that allowed user to view and analyse data without leaving the page. These include: a dynamic introgression heatmap, tree navigation of accessions, and region-focused querying of genomic variants.

The field of tomato genomics, however, is growing with increasing emphasis on mapping introgressions and higher volumes of genomic data. As the field has built on foundational work of Aflitos et al, (2014)\cite{Aflitos2014}, which sequenced 84 tomato accessions and wild relatives, exposed severe genetic bottlenecks in cultivated tomatoes and laid the groundwork for modern introgression analysis, there is increasing requirements for more powerful and more capable introgression visualization and comparative analysis tools. 

Recent studies have expanded on this base with increasingly sophisticated tools and datasets. Notably, Liu et al. (2023)\cite{liu_effects_2023} used tomato pan-genomes built from hundreds of accessions to uncover regions of structural variation and introgression hotspots, many of which affect key traits like fruit size, shape, disease resistance, and flavour. The construction of these pan-genomes enables researchers to observe variation absent in the reference genome, improving trait mapping and introgression detection.

Here, we present Tersect Browser+, an updated introgression analysis platform with expanded front-end features extension capabilities that users can view and analyze data without leaving the page.  It integrates JBrowse (Skinner et al., 2009\cite{skinner_jbrowse_2009}; Buels et al., 2016 \cite{buels_jbrowse_2016}) capabilities to capture gene models, exon-intron structures, and positional gene information for biologists inspecting tomato introgression data sets. 

JBrowse is a widely used, web-based genome browser designed for fast, interactive visualization of genomic data. Developed as a successor to earlier tools like GBrowse, JBrowse enables users to view and explore large-scale genome annotations, sequence data, and variation information through a modern, dynamic interface (Skinner et al, 2009 \cite{skinner_jbrowse_2009}). It supports a wide variety of standard file formats, including GFF, BAM, VCF, and FASTA, and is highly extensible through plugins, making it adaptable for a range of research applications from comparative genomics to functional annotation, including introgression web-app based analysis (Buels et al., 2016\cite{buels_jbrowse_2016}). Researchers often deploy JBrowse to visualize genome assemblies, transcriptomes, and variant information. Its active development and ease of integration into web servers have made it a common choice for bioinformatics platforms. 

Tersect Browser+ also includes other features such as a search feature and barcode generating feature. It allows for the investigation of other crop datasets, for example Soybean [include specific species type/name]. We identify previously published introgressions with our tool, illustrating its potential for future research, and enriching plant breeders' understanding of crop cultivar genomics.

% EARLIER TEXT KEEP UNTIL SAFE TO DELETE
%\section{Expanding landscape of plant genomic studies}\label{sec2}
% As previously shown by Kurowski & Mohareb (2020), Tersect, a lightweight, high-performing, real-time genome querying command-line utility that uses its own query language and fast execution through bitmap indices to interpret and apply flexible set theortical expressions to sets of sequence variant data, allows [something] identification of key variants and structures like haplotypes and introgressions in support of comparative genomic studies. Tersect returned  
% The field of tomato genomics has continued to evolve rapidly, with a growing emphasis on mapping introgressions â€” segments of DNA from wild tomato species that have been incorporated into cultivated lines. Foundational work such as that by Aflitos et al. \cite{RN7} , which sequenced 84 tomato accessions and wild relatives, exposed severe genetic bottlenecks in cultivated tomatoes and laid the groundwork for modern introgression analysis. 
% Recent studies have expanded on this base with increasingly sophisticated tools and datasets. Notably, Liu et al. (2023) used tomato pan-genomes built from hundreds of accessions to uncover regions of structural variation and introgression hotspots, many of which affect key traits like fruit size, shape, disease resistance, and flavour. The construction of these pan-genomes enables researchers to observe variation absent in the reference genome, improving trait mapping and introgression detection. 
% \subsection{Background on introgression studies}\label{subsec1}
% The role of introgression in water-tolerance Kubond et al. (2022), shows how such analysis will be important in identifying variant resistance to climate change.  
% To manage and interpret the expanding complexity of genomic data, a wide range of bioinformatic tools have been deployed. Established platforms such as SnpEff, VEP, BEDTools, BEDOPS, CrossMap, and the UCSC Genome Browser continue to support annotation, variant effect prediction, and coordinate translation. Visualization utilities such as the Introgression Browser (Aflitos et al., 2015) and the Tersect Browser (Kurowski and Mohareb, 2019) offer tailored environments for exploring introgression patterns, with Tersectâ€™s set-theoretic framework enabling high-speed, complex genomic queries. 
% Meanwhile, long-read sequencing technologies (e.g., PacBio, Hi-C, and optical mapping) as applied in SL4.0 (Hosmani et al., 2019), and innovations in in situ hybridization (Shearer et al., 2014), have markedly improved the resolution of reference genomesâ€”critical for detecting subtle and ancient introgressions. 
% Beyond tomato, research in soybean introgression offers valuable cross-crop insights. Studies have shown how introgressed genomic regions from wild Glycine soja into Glycine max improve traits like drought resistance, seed oil composition, and disease resilience. Tools such as SoyFGB v2.0 \footnote{\url{https://sfgb.rmbreeding.cn/index}}, SoyBase\footnote{\url{https://www.soybase.org/}}, and SoyKB\footnote{\url{http://soykb.org/}} provide functional genomics platforms that may serve as models for tomato-specific resources. Graph genome approaches and machine learning models are emerging as promising directions for future introgression prediction and trait association.
% Together, these advances paint a dynamic picture of tomato genomics: one where introgression is not only a window into domestication and evolutionary history but also a powerful mechanism for ongoing improvement, including against threats such as climate change. The fusion of high-resolution genomic data, computational methods, and specialized tools enables researchers to explore tomatoâ€™s genetic landscape with increasing clarity and sophistication.  % (Introgression hybridisation is an important factor in crop improvement. Biologists need to determine introgressions and identify donor species through visualising genetic distance and phylogenetic relationships based on the whole genome variant data. However, existing software is not suitable for full exploitation of the large publicly available data sets. Tersect Browser is a Web Application that is optimised for generating such visualisations.)

\section{Features}

% # Built based on Tersect Browswer... About Tersect About Tersect+- unified database, indexes and tracks and how speed is achieved. 
% Builds a unified allele database from VCF files, using bitmap indexing to track presence/absence of variants. The system links variant positions to bitmap indices, allowing efficient region-specific searches and genome comparisons via compressed, sparse data structures. Indexing is fast and optimized for read-heavy workflows. While updating is less efficient, batch additions mitigate this. A custom parser converts set-based queries into abstract syntax trees (ASTs), which are then optimized and reused across queries for speed.  We summarize here the functionality. Probably in the following structure. Browser; Models; Search; and Barcodes and Introgressions [if we get there.] We do this with a paragraph on each. We write similarly to the Fady and Tom email. We could do it in two parts:  Browser and Gene Model and Search and Barcodes including introgression search, we do that. The structure in paragraphs will be:

% Browser  - all these will be kept short paragraphs.
% Gene Models
% Search 
% Barcodes
% [Introgression]

Tersect Browser+ is a web application that depends on resequenced data in VCF format and uses a reference genome in FASTA format. Utilizing a MEAN stack (MongoDB, Express, Angular, and Node) architecture, it creates a webpage displaying heatmaps and phylogenetic trees for introgressions within the loaded crop dataset. This provides an interface for users to identify differences across accessions within the same species dataset. It also includes a JBrowse interface for investigating specific variants, displaying selected genes directly alongside the generated heatmap. A search feature highlights high-impact variants on the webpage, helping to distinguish the selected reference genomes. Tersect Browser+ also offers barcode generation functionality for various cultivars and specific species in the dataset.

\begin{figure} [http]
  \centering
  \includegraphics[width=0.5\textwidth]{path/to_image.png}
  \caption{Tersect Browser+ overview. The browser provides a single way to analyse variant data without leaving the page.[More description]}
  \label{fig:your_label}
\end{figure}

\subsubsection{Heatmap and Phylogenetic Tree}

Visualising introgression distances for each accession against the reference genome provides cultivar biologists with insights into genetic similarity, divergence patterns, and potential introgressed regions that may influence desirable traits. Tersect Browser displayed a phylogenetic tree and corresponding heatmap for the selected reference genome, chromosome, chromosomal interval, and bin size. The heatmap represented the genetic distances between each accession and the reference genome (Kurowski, 2023).
Tersect Browser+ extends and enhances these capabilities

\subsubsection{Gene Browser and Model}

Making it easier for users to explore gene models and genetic variants for a specific accession, this feature provides two ways to view the gene model track and variant track for the selected accession.  The user can either select the â€˜View Gene-Modelâ€™ button to load these tracks above the accession heatmap [Ref Technical Document Figure] or right-click on a heatmap region of interest to display the accession name and open the JBrowse interface [Add reference figure from the Technical document].  This feature can also be used to [Add any other information on how this can be used].
Gene model annotations are shown in the track [Add Technical Documentation figure reference]. The paths are sorted and a compressed GFF file, along with its corresponding Tabix index are specified to the local server URLs during the browser start-up.  


\subsubsection{Search Features}

High-impact variants, such as frameshift mutations, premature stop codons, or major missense mutations, can have profound effects on gene function. In crop breeding programs leveraging introgression, these variants are particularly important for assessing the functional consequences of transferred genomic regions. By enabling users to search for high-impact variants based on gene ID and impact severity, Tersect Browser+ facilitates the identification of critical genetic changes associated with desirable or undesirable traits. Selected variants are visually mapped onto the heatmap, highlighting the corresponding accession and genomic region [Figure from Technical Document reference].

\subsubsection{Barcode Generator}

Providing a DNA barcode or fingerprint of the genomic composition of each accession enables rapid identification of regions derived from donor versus recipient genomes. This allows breeders to verify the successful integration of target traits and to track the use of parental lines, which is useful in identifying specific accessions in the laboratory or published datasets. In recent years, significant advances have been made in DNA barcoding methodologies. For example, Prieto et al. (2024)\cite{espinosa_prieto_finding_2024} demonstrated its application in DNA fingerprinting of tomato breeding lines, while Al-Shammari et al. (2021) \cite{al2021genetic} documented its use in cultivar investigations. In Tersect Browser+, users can generate barcodes of their preferred length for any region on a selected chromosome. The tool identifies key characteristics of interest within the barcode region, including SNP counts and positions, the number of repeat elements, and GC content [Ref figure in technical documentation showing output]. A default barcode length of 150 bp is provided, reflecting the attractiveness of lengths between 150â€“500 bp for primer production, as discussed by Letsiou et al. (2024) \cite{letsiou_dna_2024}.

\subsubsection{Datasets}

The Sol Genomics Network (SGN) tomato reference genome and associated resequenced accessions from cultivated lines and wild relatives was used. Reference genome updates (e.g., SL2.40 to SL2.50) were managed using SeqRemap, available under the MIT license at \url{https://bitbucket.org/cranfieldbix/seqremap}.

\vspace{3mm}

\noindent[Add on SoyBean]

\section{Funding}

[Tom - we see that funding sections are covered in many papers submitted to this journal. Do we need to include? If so, what do we need to include in here?]

\vspace{3mm}

\noindent\textit{Conflicts of interest}:~none declared.

\section{Supplementary Data}

Technical documentation - pdf file 

\noindent\url{[githubreadme - to be added]}

% \section{Competing interests}

% \section{Author contributions statement}

% \section{Acknowledgments}

\bibliographystyle{plain}
\bibliography{Article/article_references}


\end{document}
