\documentclass[12pt]{article}
\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{xcolor}
\usepackage{hyperref}
\usepackage{listings}
\usepackage{soul}
\usepackage[a4paper,width=150mm,top=25mm,bottom=25mm]{geometry}

\title{
{\includegraphics[width=3cm, height=2.5cm]{Cran.jpg}}
\\
\includegraphics[width=7cm, height=1cm]{TB.jpg}
\\
{User Manual for TersectBrowser+}
}



\author{David Oluwasusi, Tanya Stead, Gregory Lupton, Gabrielle Baumberg \\ Supervised by: Dr. Tomasz Kurowski}

\begin{document}
\sloppy % allows line spacing stretch to justify the text

\maketitle

\section{Overview}
TersectBrowser+ is designed to be useful to a range of Users. Hosting features within the same browser allows the conducting of full analyses within the same page, benefitting both a more experienced researcher as well as a novice to the field. 

\section{Use Cases - Video transcript}
1a. TB Homepage & main page
On opening of TersectBrowser+, we are met with the introductory page displaying the loaded datasets. The user can click on one of the available datasets, which will bring them to the main TersectBrowser+ exploratory page. The main body of the page is composed of a heatmap, comparing all accessions within the dataset against the selected reference accession. Darker bins represent areas with greater variation between the accession and the reference, whilst lighter bins represent more similar sequences. Vertical red bars signify gaps in the reference genome. Above the heatmap is a scale bar indicating chromosome position. 

1b. TB Main page header and buttons
The menu bar above the heatmap contains navigational controls. Newly added in the TersectBrowser+ release is the `binoculars' button, the `Open Variant Search' button, and the `refresh' button.

2a. Genome browser main window
To view gene models corresponding to chromosome position, the user can click on the `binoculars' button, which will open the genome browser panel above the heatmap, showing gene models for the corresponding chromosomal positions. The panel has a left-margin offset to ensure the chromosomal scale bar between the genome browser panel and heatmap are aligned. In this view, the genome browser panel and heatmap are fully synced, such that any changes made to the horizontal scroll along the x-axis, zoom level,bin size, selected interval, or selected chromosome, will be immediately reflected in both the heatmap and genome browser panel. 

The initial view will show an error message, prompting the user to zoom in on decrease the bin size, as the default view of TersectBrowser+ is too zoomed out. The genome browser panel has a set height that can be changed by clicking and dragging the bottom of the panel. The user can also scroll vertically along the y-axis to view all gene models without changing the view of the TersectBrowser+ heatmap. 

2b. Genome Browser popup window
If the user is interested in a specific accession or a specific bin, they can click on the specific accession or bin, which will open a popup window displaying a `View in browser' button. Clicking on this button will open a large popup window that is detached from the background TersectBrowser+, allowing the user to change the view by zooming or scrolling in the popup window without changing the view in TersectBrowser+. If the user has opened the genome browser popup window from accession name, the window will be fully zoomed out and show the entire chromosome. If the user has opened the genome browser popup window from a specific bin, the window will be zoomed in to show that particular interval. 

3. Feature Search
The Feature Search functionality enables the user to search for specific variants (High/Moderate/Low) present within all accessions across a specified interval or within a specific gene. Clicking on the `Open Variant Search' button opens a popup window where the user can specify an interval to search, either by using the sliding bar or by manually typing in numbers, and can specify an impact level from the dropdown menu. Alternatively to searching by interval, the user can search by gene ID. The gene search bar has an autocomplete functionality and will suggest all possible gene IDs alphabetically. Searching by interval and gene ID are mutually exclusive, and when the user starts to type a gene ID the interval functionality becomes greyed out. To initiate the variant search, the user clicks on the `Search for variants' button, which highlights bins in the canvas representing the positions and accessions where variants of the specified impact level are present.

Additionally, the user can select a specific bin or click and drag the chromosome scale bar to select an interval, which will bring up a popup menu with a `Search for variants' button. Clicking on this button will automatically open the `Variant Search' popup window with the interval preselected, and will automatically initiate the variant search. Again, bins in the heatmap will be highlighted red if they contain a variant of the specified impact level.

4. Barcode generation
Once the user has identified an accession and a bin of interest, TersectBrowser+ offers the capability to generate accession-specific barcodes for that interval. Clicking on a specific bin opens a popup menu with a `Create barcode' button. Clicking this button opens a larger Barcode Generator popup window, where the user can specify barcode size and the maximum number of variants per barcode, which are populated with the default values on 150 and 1 respectively. Clicking on the `Generate' button will create the barcodes, and a `Download' button will appear. Clicking the `Download' button automatically downloads a .tsv file containing the generated barcodes and their associated statistics.

5. Licensing and contact


\subsection{Example 1:} A tomato genetics researcher wishes to identify differences between a domestic tomato genome and a large number of resequenced genomes of wild relatives. 

\paragraph{User profile:} Extensive knowledge of introgression genetics and in particular the tomato genome and different species. Lacks knowledge of software development and the different file types involved in setting up the software. 

\begin{enumerate}
  \item The researcher reaches out to administration or informatics support to set up an instance of TersectBrowser+ that contains the large resequenced dataset and relevant GTF files, Reference genomes, and Metadata (e.g. geographic location of species) annotations for the difference accessions. 
  \item The researcher uses a Local Host Web Address to access TersectBrowser+. The default view is a list of all accessions in the dataset, with one chromosome view selected and a set bin size and zoom level. 
  \subitem  At this stage, the researcher can see a heatmap of SNP density across the different accessions, and gain a brief overview of which accessions differ most from the selected reference sequence. The researcher also can see which locations on the chromosome contain the most variation in SNP density. 
  \item The researcher is particularly interested in a dark section of the heatmap (I.e. high density of SNPs relative to the selected reference accession), that lies across a chromosomal region containing genes involved in heat stress regulation. The researcher uses the click and select feature along the scale bar to bring up a popup message with further options, and selects the "Set as interval" feature. 
  \subitem At this stage the researcher understands that the view will change so that the selected chromosomal region (containing high density SNPs in certain accessions) will be zoomed to in some way.  
  \subitem The researcher sees that TersectBrowser+ view zooms to the selected region. The variation in grey bin highlighting is more obvious in this view. The accession names are also zoomed in, and therefore easier to read.
  \item The researcher now sees that a particular set of 5 accessions are the darkest on the heatmap. The researcher would like to know if any key genes cover the dark regions for these 5 accessions. They understand that the gene browser containing gene models and the reference genome is accessible by clicking the binoculars button.  
  \subitem The researcher sees a new view with the heatmap shifted down so there is a panel of the gene browser at the top of the page. The default view of the gene browser panel is the gene models track, and the first two accession vcf files in the resequenced dataset. The researcher clicks the track selecter button to search for the names of the 5 particular accessions. 
  \subitem The researcher finds that the zoom on the heatmap has a minimum level past which the scroll zoom no longer works. However, the gene browser panel can be zoomed further than this by selecting a region on the scale bar and clicking "Zoom to region". 
  \subitem This then zooms to a view where the gene models are visible alongside the variants specific to each accession.  
  \item The researcher finds a gene model that they know to be important for heat stress. Upon clicking this gene in the browser \hl{panel,}
\end{enumerate}

\subsection{Example 2:} A plant breeder intends to gain intellectual property (IP) recognition for a particular trait of their Soybean plant breed, and wants to find quantitative evidence to support their application using the genome of this breed. 

\paragraph{User profile:} High familiarity with phenotypic traits of different Soybean plant breeds, as well as history of intended Soybean introgressions by different plant breeders. Lacks knowledge of statistical generation of phenetic trees using pairwise relationships, as well as how the data in the TersectBrowser+ is set up to view the whole dataset. 

The breeder opens up the phenetic tree view, in order to view a summary of pairwise differences between the different accessions. 

The breeder uses the barcode search function to identify whether the sequence within the identified introgression in their plant breed can be used to uniquely identify that breed relative to the rest of the resequenced dataset. 

The breeder exports the results of the barcode search to use in their application for IP recognition of their breed. 

\subsection{Example 3:} A bioinformatics expert intends to learn more about what introgression means in practice as relating to differences between genomes of similar species. 

\paragraph{User profile:} Expert knowledge in GUI Web-app design, as well as  

The bioinformatician can zoom in the view using the plus/minus buttons on the home bar, or using the mouse zoom scroll (up and down). 

The bioinformatician uses the introgression prediction tool to identify which high SNP density regions meet the threshold requirements for high likelihood of introgression from a wild cultivar into the domestic cultivar. 

\section{Admin setup of browser}
The Admin is defined as an individual with bioinformatics and software design knowledge, background comprehension for all sections of this Technical Documentation, and acting to facilitate the use of a deployed TersectBrowser+ by the plant breeder main user of the software. When a new dataset is requested for deployment on TersectBrowser+, the admin will need to follow the below steps.

\subsection{Download of required material}
\begin{enumerate}
  \item Collect background context and required data from the user.
\begin{itemize}
  \item Reference genome in fasta format (may be compressed). Size and chromosome number will impact the speed of dataset upload and deployment - TersectBrowser+ has been tested with a genome size of ~1GB, and 20 chromosome pairs.
  \item Resequenced genome dataset in VCF format (may be compressed, as a multi-sample VCF or as a directory of individual files). TersectBrowser+ has been tested with a dataset size of ~500 VCF files representing individual accessions. 
  \item Any metadata associated with different accessions in the dataset, such as wild variety vs domesticated variety. \hl{This should be provided in a text file.}
  \item A GFF file of gene model information for the reference genome, produced using SNPeff software. If not using SNPeff, ensure that variant impact levels are categorised into 'High', 'Medium', and 'Low' impact. If not provided, the gene model track in the Variant Browser will be unavailable. 
  \item Whether there are prior identified introgressions known to be functionally relevant for the species. These can be used to 'check' the ability of TersectBrowser+ to identify introgressed regions for the specified dataset.
\end{itemize}
  \item Clone the GitHub repository of TersectBrowser+ to the local machine. Follow the README from \hyperlink {https://github.com/Tersect-Browser/Tersect-browser.git}{TersectBrowser+ GitHub} for installation instructions.
  \item Download relevant files from section (1) to a new folder within the root directory of the local Tersect-browser repository.
  \item Set environment variables \verb+fasta+, \verb+gff+, and \verb+vcfs+ according to their new location relative to the root of Tersect-browser.
  \item Run the script \verb+setup_new_tbrowser_dataset.py+ using the environment variables of files as input:
  
      \verb+python setup_new_tbrowser_dataset.py -f ${fasta} -g ${gff} -V ${vcfs}+
      \begin{itemize}
          \item After running the script, there should be visible a new \verb+config.json+ file in the root folder, as well as a new folder \verb+./~/mongo-data/gp_data_copy+ containing fasta files, gff files, and VCF files along with their index files. The main Tersect index file should be in the folder above. This is where the server will look to find information on tracks when generating the Variant Browser panel.
      \end{itemize}
\end{enumerate}

\section{Config file generation}
There are multiple extensions incorporated within TersectBrowser+, and all config files must be correctly generated for full functionality of the browser. However, each extension has different config requirements and may allow limited functionality before the generation of all config files.

\subsection{Variant Browser Pane}
A config.json file containing a/the genome assembly is first generated. 
\paragraph{Using JBrowse CLI}\mbox{}
\\\\
The fasta file must first be indexed as follows (this step can be omitted if a .fai file already exists):
\lstset{
  language=bash,
  basicstyle=\ttfamily
}
\begin{lstlisting}
    $ cd Tersect-browser/~mongo-data/gp_data_copy/
    $ samtools faidx genome.fa 
\end{lstlisting}
The config.json file is created and the genome assembly added to it via the following command:
\begin{lstlisting}
    $ jbrowse add-assembly <genome.fa> --load copy --out config.json
\end{lstlisting}
The variant tracks can then be added to the \verb+config.json+ file via the following commands. Jbrowse CLI expects the VCFs to be compressed with bgzip and indexed using tabix. 
\begin{lstlisting}
    $ bgzip file.vcf 
    $ tabix file.vcf.gz 
    $ jbrowse add-track file.vcf.gz --load copy --out config.json 
\end{lstlisting}
The generated config file will contain both the \verb+assembly[]+ and the \verb+tracks[]+. These can then be manually copied and pasted into their respective assembly.ts and tracks.ts files - the reference genome fasta will be added to the \textit{extension/genome-browser/src/app/react-components/assembly.ts} file, while the GFF file and all accession VCF tracks will be added to the \textit{extension/genome-browser/src/app/react-components/tracks.ts} file.

Correct config specifications for the reference genome assembly can be found \textunderscore{\hyperlink{https://jbrowse.org/jb2/docs/config_guides/assemblies/}{here}}. 

\paragraph{Manual Generation}\mbox{}
\\\\
If Jbrowse CLI is not compatible with the machine, the assembly config can be generated by adding the following text, and \hl{changing the parameters trackID, name, and URI where necessary}. 

\begin{lstlisting}
    $ cd extension/genome-browser/src/app/react-components/
    $ nano config.json

    "name": "SL2.50", 
    "sequence": { 
      "type": "ReferenceSequenceTrack", 
      "trackId": "SL2.50-ReferenceSequenceTrack", 
      "adapter": { 
        "type": "IndexedFastaAdapter", 
        "fastaLocation": { 
          "uri": "http://localhost:4200/TersectBrowserGP/tbapi/datafiles//SL2.50.fa", 
          "locationType": "UriLocation" 
        }, 
        "faiLocation": { 
          "uri": "http://localhost:4200/TersectBrowserGP/tbapi/datafiles//SL2.50.fa.fai", 
          "locationType": "UriLocation" 
        } } } 
\end{lstlisting}

The track config can also be generated by adding the following text, and \hl{changing the parameters trackID, name, and uri where necessary}. 

\begin{lstlisting}
    { 
    "type": "VariantTrack", 
    "trackId": "S.lyc LA2706", 
    "name": "S.lyc LA2706", 
    "adapter": { 
        "type": "VcfTabixAdapter", 
        "vcfGzLocation": { 
            "uri": "http://localhost:4200/TersectBrowserGP/tbapi/datafiles/RF_001_SZAXPI008746-45.vcf.gz.snpeff.vcf.gz", 
            "locationType": "UriLocation" 
        }, 
        "index": { 
            "location": { 
                "uri": "http://localhost:4200/TersectBrowserGP/tbapi/datafiles/RF_001_SZAXPI008746-45.vcf.gz.snpeff.vcf.gz.tbi", 
                "locationType": "UriLocation" 
            }, 
            "indexType": "TBI" 
        } 
    }, 
    "assemblyNames": [ 
        "SL2.50" 
    ]      } 
\end{lstlisting}

A script to automate this process (the tracks will be more numerous than assemblies) can be found on \hl{Elvis} at \textit{gp\_data\_copy/scripts/generate\_track\_config.py}, and  a script to generate tbi files for all VCF files, \textit{generate\_tbi.sh}, is also present in this folder. 

\subsection{Gene Model Track}
\paragraph{Using JBrowse CLI}\mbox{}
\\\\
Once a \verb+config.json+ file for the main reference assembly has been created, run the following commands to sort the GFF3 file, create the .tbi file, and add the config files to the \verb+config.json+ using these Jbrowse commands:
\lstset{
  language=bash,
  basicstyle=\ttfamily
}
\begin{lstlisting}
    $ jbrowse sort-gff <yourfile.gff> | 
        bgzip > <yourfile.sorted.gff.gz>
    $ tabix <yourfile.sorted.gff.gz>
    $ jbrowse add-track <yourfile.sorted.gff.gz> --load copy
\end{lstlisting}
This will create a \verb+FeatureTrack+ under the \verb+tracks[]+ section in the \verb+config.json+. Then copy and paste the \verb+FeatureTrack+ to the \textit{extension/genome-browser/src/app/react-components/tracks.ts} file.


Note: the Jbrowse "sort-gff" command just automates the following shell command 
\begin{lstlisting}
    $ (grep "^#" in.gff; grep -v "^#" in.gff | 
        sort -t"`printf '\t'`" -k1,1 -k4,4n)  
        > sorted.gff; 
\end{lstlisting}

\paragraph{Manual Generation}\mbox{}
\\\\
If Jbrowse CLI cannot be accessed, the \verb+FeatureTrack+ can be generated by adding the following text, replacing trackID, name, and uri with the respective inputs:
\begin{lstlisting}
    { 
        "type": "FeatureTrack", 
        "trackId": "ITAG2.4 Gene Models", 
        "name": "ITAG2.4 Gene Models", 
        "adapter": { 
          "type": "Gff3TabixAdapter", 
          "gffGzLocation": { 
            "uri": "http://localhost:4200/TersectBrowserGP/tbapi/datafiles/ITAG2.4_gene_models.sorted.gff3.gz", 
            "locationType": "UriLocation" 
          }, 
          "index": { 
            "location": { 
              "uri": "http://localhost:4200/TersectBrowserGP/tbapi/datafiles/ITAG2.4_gene_models.sorted.gff3.gz.tbi", 
              "locationType": "UriLocation" 
            }, 
            "indexType": "TBI" 
          } 
        }, 
        "assemblyNames": [ 
          "SL2.50" 
        ] } 
\end{lstlisting}

\subsection{Feature Search}



\subsection{Barcode Generation}


\section{User access of browser}




\end{document}